enclave-runner/src/tcs.rs

```diff
@@ -54,31 +57,145 @@ pub(crate) fn coenter<T: Tcs>(
+                // of the unconditional jump above.
+.global __vdso_clock_gettime
+                call __vdso_clock_gettime@PLT
+
+1:
+                ", out(reg) addr, options(nomem, nostack, att_syntax));
+            addr != 0
+        }
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    fn has_vdso_sgx_enter_enclave() -> bool {
+        false
+    }
+
     let sgx_result: u32;
-    let mut _tmp1: u64;
-    let mut _tmp2: u64;
 
     unsafe {
         let mut uninit_debug_buf: std::mem::MaybeUninit<DebugBuffer>;
-        let debug_buf = debug_buf.map(|r| r.borrow_mut());
-        let debug_buf = match debug_buf {
-            Some(mut buf) => buf.as_mut_ptr(),
+        let mut debug_buf = debug_buf.map(|r| r.borrow_mut());
+        let debug_buf_ptr = match debug_buf {
+            Some(ref mut buf) => buf.as_mut_ptr(),
             None => {
                 uninit_debug_buf = std::mem::MaybeUninit::uninit();
                 uninit_debug_buf.as_mut_ptr() as *mut _
             }
         };
-        llvm_asm!("
-        lea 1f(%rip),%rcx
-1:
-        enclu
-"       : "={eax}"(sgx_result), "={rbx}"(_tmp1), "={r10}"(_tmp2),
-              "={rdi}"(p1), "={rsi}"(p2), "={rdx}"(p3), "={r8}"(p4), "={r9}"(p5)
-            : "{eax}" (2), "{rbx}"(tcs.address()), "{r10}"(debug_buf),
-              "{rdi}"(p1), "{rsi}"(p2), "{rdx}"(p3), "{r8}"(p4), "{r9}"(p5)
-            : "rcx", "r11", "memory"
-            : "volatile"
-        )
+        if has_vdso_sgx_enter_enclave() {
+            #[repr(C)]
+            #[derive(Default)]
+            struct SgxEnclaveRun {
+                tcs: u64,
+                function: u32,
+                exception_vector: u16,
+                exception_error_code: u16,
+                exception_addr: u64,
+                user_handler: u64,
+                user_data: u64,
+                reserved: [u64; 27],
+            };
+
+            impl fmt::Debug for SgxEnclaveRun {
+                fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                    let function = Enclu::try_from(self.function);
+                    let fundbg = match &function {
+                        Ok(v) => v as &dyn fmt::Debug,
+                        Err(_) => &self.function,
+                    };
+                    f.debug_struct("SgxEnclaveRun")
+                        .field("function", fundbg)
+                        .field("exception_vector", &self.exception_vector)
+                        .field("exception_error_code", &self.exception_error_code)
+                        .field("exception_addr", &(self.exception_addr as *mut ()))
+                        .finish()
+                }
+            }
+
+            let mut run = SgxEnclaveRun {
+                tcs: tcs.address() as _,
+                ..Default::default()
+            };
+            let ret: i32;
+            asm!("
+                    sub $0x8, %rsp                   // align stack
```

Comment:
Why is this required?

Reply:
Call stack alignment is discussed in section 3.2.2 of the [ABI](https://github.com/hjl-tools/x86-psABI/wiki/x86-64-psABI-1.0.pdf). According to [`asm!` documentation](https://doc.rust-lang.org/beta/unstable-book/library-features/asm.html#options-1):

> If [the `nostack`] option is not used then the stack pointer is guaranteed to be suitably aligned (according to the target ABI) for a function call.

However, we are passing a single argument on the stack, so we need to make sure the stack is properly aligned after the stack arguments.

Reply:
Oh I didn't know about that requirement. Good to know!